### ISSUE TITLE
[Questions] Little reward (<0.1%) when training inliner using LTO

### ISSUE THREAD

**Issue by DataCorrupted** [google/ml-compiler-opt#490]
Hi all, first off huge thanks for the great work you published!

We are trying to train an inliner model for our use. The first step we did is try to train a model based on lld's corpus (~1600 objects)

# TL;DR
We can see great results when training on no LTO,  p50 of reward per object is around 10%, and after linking we can get 0.5% wins (we turned on many linker optimizations)
However, when training on thin LTO, we can see little to no reward. p50 is zero and mean is <0.1%. After analysis that mean reward comes from some small files (<1kb) getting 5 to 10% rewards while the majority files didn't change at all.

Here's a mean reward we cropped from our tensorboard: 
<img width="1049" alt="Image" src="https://github.com/user-attachments/assets/2630ad31-4499-4a62-bf9f-44753ef97ad9" />

We did some investigation, and find that the LTO objects initial size is generally smaller than ML optimized object size, indicating that there may be nothing left to inline. But I don't have enough evidence to draw the conclusion.

# Questions
I see there is a doc for [thin LTO training](https://github.com/google/ml-compiler-opt/blob/main/docs/thinlto.md), can we enrich that? 
More specifically, I wonder what does the build/collect trace process look like when training Chrome. What is the reward growth per object and what is the final binary win?

# Detailed configs

Here's a detailed configurations we used for training, in case it matters:

```sh
# build LLVM
FLAGS="-isysroot ... \
    -Oz -Wl,-O2 
    -target arm64-apple-darwin \
    -fuse-ld=lld
"
case "$LTO" in
    thin)
        echo "LTO is set to thin LTO"
        # Add any specific commands or flags for thin LTO here
        FLAGS="$FLAGS -flto=$LTO -Wl,-save-temps -Wl,--thinlto-emit-index-files"
        ;;
    no)
        echo "LTO is set to no LTO"
        # Add any specific commands or flags for no LTO here
        FLAGS="$FLAGS -Xclang=-fembed-bitcode=all"
        ;;
esac

cmake -G Ninja  \
    -S $LLVM_SOURCE \
    -B $LLVM_TRAIN_XBUILDDIR \
    -DCMAKE_BUILD_WITH_INSTALL_RPATH=ON  \
    -DCMAKE_C_COMPILER="clang" \
    -DCMAKE_CXX_COMPILER="clang++" \
    -DCMAKE_LIBTOOL="llvm-libtool-darwin" \
    -DLLVM_TABLEGEN="llvm-tblgen" \
    -DCMAKE_BUILD_TYPE=Release \
    -DLLVM_DEFAULT_TARGET_TRIPLE=arm64-apple-darwin \
    -DLLVM_ENABLE_PROJECTS='llvm;lld' \
    -DLLVM_TARGET_ARCH='AArch64' \
    -DLLVM_TARGETS_TO_BUILD='AArch64' \
    -DLLVM_BUILTIN_TARGETS='arm64-apple-darwin' \
    -DLLVM_RUNTIME_TARGETS='arm64-apple-darwin' \
    -DCMAKE_CXX_FLAGS_RELEASE='-Oz' \
    -DCMAKE_C_FLAGS_RELEASE='-Oz' \
    -DCMAKE_CXX_FLAGS="$FLAGS" \
    -DCMAKE_C_FLAGS="$FLAGS"
ninja -C $LLVM_TRAIN_XBUILDDIR lld
```

```sh
# Extract corpus
# Base command for extract_ir.py
EXTRACT_IR_CMD=(
  python3 llvm-project/llvm/utils/mlgo-utils/mlgo/corpus/extract_ir.py
  --cmd_filter="^-O3|-O2|-Os|-Oz$"
  --input_type=json
  --llvm_objcopy_path="llvm-objcopy"
  --output_dir="$INL_CORPUS"
  --cmd_section_name="__LLVM,__cmdline"
  --bitcode_section_name="__LLVM,__bitcode"
)
# Add thinlto_build option if LTO is thin
if [ "$LTO" == "thin" ]; then
    EXTRACT_IR_CMD+=(
      --obj_base_dir="$LLVM_TRAIN_XBUILDDIR"
      --thinlto_build=local
    )
else
    EXTRACT_IR_CMD+=(
      --input="$LLVM_TRAIN_XBUILDDIR/compile_commands.json"
    )
fi

"${EXTRACT_IR_CMD[@]}"
```

```sh
# Override for thinLTO
    NEW_OVERRIDE=(
        "-isysroot"
        "..."
        "-Oz"
        "-target"
        "arm64-apple-darwin"
        "-c"
    )
```

Extract trace and training are the same as documentations.

**Comment by mtrofin** [comment]
I'm assuming / questions (I'll number them, makes things easier :) ):

1. you're collecting a new vocab after collecting the thinlto corpus
2. also, the linker optimizations are the same in both no thinlto and with thinlto cases
3. you're `strip -s` -ing the final binary (either that, or we're looking at `llvm-size` output for `.text`)
4. are you applying the ml advisor both in the front and backend(==post-thinlink), when evaluating?
5. Training chases the .text size as reported by `llvm-size` so it's possible there are second order effects that hide those savings. Are you seeing a difference in savings when looking just at .text vs when looking at file size?

Anecdotally, using an off-the-shelf model[1], I'm also seeing very little improvement in lld, well, specifically:
- stripped files, under `ls -l`: almost no difference (0.001%)
- same files, `.text` section as reported by `llvm-size`: 0.23%


Re. Chrome: the most recent set up and data is https://chromium-review.googlesource.com/c/chromium/src/+/5784535?checksPatchset=4&tab=checks. The size report is given by the APK Size check, but you have to expand the "Info" and then click on "APK Size" to get to the 64-bit outcome the patch is about. I don't have a per-module breakdown.

Note that that's with skipping over hot callers, which we inline using the default, -O2 policy. Also this was with the imitation learning work @tvmarino has been porting, but IIRC that is on top of the "normal" training, and while it should improve things, I don't recall it doing that to explain a gap like you're getting.

I don't have ready a per-module breakdown, @tvmarino do you happen to have that from your immitation learning work?

[1] which we forgot to release, doing that shortly. It should be not too far off the "
inlining-Oz-v1.1" one, though.

**Comment by DataCorrupted** [comment]
1. Yes
2. Yes
3. We tried both, currently we rely on `llvm-strip`. But it should not bring a huge difference
4. Interesting questions. We trained two models (NoLTO and ThinLTO), here's what we tried:
    - Apply NoLTO to pre-link only: ~0.5% size win
    - Apply ThinLTO to post-link: ~0.5 size regression
    - Apply NoLTO to pre-link on a thin LTO build: 1% regression
    - Apply ThinLTO to bost pre-link and post-link on a thin LTO build: ~3% regression
5. If you mean the difference `.text` size and stripped binary, no, I see little change

Also, thanks for the chrome example. Looking forward to the release of the new inline model.

**Comment by mtrofin** [comment]
> 5. If you mean the difference `.text` size and stripped binary, no, I see little change

To clarify: the relative changes in `.text` and the relative changes in binary size follow the same trend, correct?

**Comment by kyulee-com** [comment]
Beyond the linker optimizations that affect the final result, the issue seems somewhat fundamental even at this training level on  each module size for ThinLTO. NoLTO bitcode contains many linkonce_odr linkage functions (from template instantiation in C++), while most of them become available_externally linkage functions, except for the leading/prevailing one in ThinLTO bitcodes that become promoted after the thin-link. The former contributes to the object file size, while the latter does not. This means there is more entropy and a larger search space for the NoLTO case with these unoptimized bitcodes. In other words, there is little room for improvement inherently with the current training setting/assumption with these optimized ThinLTO bitcode files.

**Comment by DataCorrupted** [comment]
> the relative changes in .text and the relative changes in binary size follow the same trend, correct?

Yes

**Comment by DataCorrupted** [comment]
> cases like chrome (~40K modules IIRC), both for training (more examples) and for reducing size (more opportunities)

Let's say the off-the-shelf model you mentioned above is trained on Chrome, and it has seen adequate examples. That might suggest lld had little inline opportunities left.

**Comment by mtrofin** [comment]
> > cases like chrome (~40K modules IIRC), both for training (more examples) and for reducing size (more opportunities)
> 
> Let's say the off-the-shelf model you mentioned above is trained on Chrome, and it has seen adequate examples. That might suggest lld had little inline opportunities left.

I agree.

...or at least little opportunities left under the current feature modeling / training assumptions, to paraphrase @kyulee-com). I guess if you have a per-module (post-thinlink) breakdown of savings and there are improvements and regressions, theoretically the improvements alone could give an indication of what _might_ be still saved (and @tvmarino's extra training step would help there). 

But, ya, probably for lld we're looking at squeezing water out of dry stone :)
